tools/stack_decode.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python
+
+# Call addr2line as needed to resolve addresses in a stack trace,
+# de-interleaving the log lines from multiple threads if required.
+#
+# Two ways to call:
+#   1) No arguments means this script will treat stdin as log output from
+#   Envoy
+#   2) Some arguments means run a subprocess with that command line and send
+#   stderr through the script
+#
+# In each case this script will decode any backtrace log lines found and echo
+# back all non-Backtrace lines untouched.
+
+import re
+import sys
+import subprocess
+
+
+# Process the log output looking for stacktrace snippets, print them out once
+# the entire stack trace has been read.  End when EOF received.
+def decode_stacktrace_log(input_source):
+  traces = {}
+  trace_begin_re = re.compile('^(.+)\[backtrace\] Backtrace obj<(.+)> thr<(\d+)')
+  stackaddr_re = re.compile('\[backtrace\] thr<(\d+)> #(\d+) (0x[0-9a-fA-F]+)$')
+  trace_end_re = re.compile('\[backtrace\] end backtrace thread (\d+)')
+
+  # build a dictionary indexed by thread_id:
+  #   value is a tuple (log_prefix, obj_file, [list of addresses])
+  while True:
+    line = input_source.readline()
```

Comment:
If I start an Envoy like this:

tools/stack_decode.py bazel-bin/source/exe/envoy-static -c config.json

Then ctrl-C does the right thing.  It does the expected thing when the output is piped into stack_decode.py too.  I can add a catch for KeyboardInterrupt to avoid the ugly stack trace, though...